sstables/mx/reader: accommodate inexact partition indexes

michoecho opened the pull request:
Unlike the currently-used sstable index files, BTI indexes don't store the entire partition keys. They only store prefixes of decorated keys, up to the minimum length needed to differentiate a key from its neighbours in sstables. This saves space.

However, it means that a BTI index query might be off by one partition (on each end of the queried partition range) with respect to the optimal Data position.

For example, if the index stores prefixes `a`, `b`, `c`,
the index has no way to know if the first index entry after key `bb`
is `b` (which might correspond to `ba` as well as `bc`), or `c`.
So the index reader conservatively has to pick the wider Data range, and the Data reader must ignore the superfluous partitions. (And there's no way around that.)

Before this patch, the sstable reader expects the index query to return an exact (optimal) Data range. This patch adjusts the logic of the sstable reader to allow for inexact ranges.

Note: the patch is more complicated that it looks. The logic of the sstable reader was already fairly hard to follow and this adds even more flags, more weird special states and more edge cases. I think I managed to write a decent test and it did find three or four edge cases I wouldn't have noticed otherwise. I think it should cover all the added logic, but I didn't verify code coverage. (Do our scripts for that even work nowadays)? Simplification ideas are welcome.

avikivity commented on sstables/index_reader.hh:
```diff
@@ -125,9 +125,9 @@ public:
+    // Returns the key for current partition, if available in the index.
     // Precondition: partition_data_ready()
-    virtual partition_key get_partition_key() = 0;
+    virtual std::optional<partition_key> get_partition_key() = 0;
```
I'd rather see the interface removed. After upgrading sstables (or in a new cluster), this will always return a disengaged optional.

We need to think about mixed-cluster performance (and keep in mind that sstables can be un-upgraded for a long time), so if this interface is critical for performance of older indexes, we can keep it. But hard to see how that can be true.

michoecho replied:
No, it will not always return a disengaged optional.

The reason `get_partition_key()` exists is that the index (both BIG and BTI) stores partition headers (partition key and partition tombstone) for partitions which are big enough to have a clustering index, so that you can take the header from the index and jump straight to the middle of the partition in the data file, without a need for a separate read just to get the partition tombstone from the beginning of the partition in the data file.

For the partitions where `get_partition_key()` is actually useful for something, both BIG and BTI return an engaged optional. But BIG returns an engaged optional even when `get_partition_key()` isn't useful (because there is clustering index, and thus no partition tombstone, and you still have to read the header in the data file to find the partition tombstone to do anything with the partition), while BTI only returns an engaged optional when the partition key is useful.

michoecho replied:
@nyh get_partition_key() (i.e. storing a full partition key in the index) is an optimization which allows the data reader to avoid reading the partition header from the Data file just to get the key, in cases where the queried partition slice is potentially far away from the header on the disk.

The reader shouldn't rely on the full key being present. That's an arbitrary choice of the writer of the index. If the full key is not present, the reader should just fall back to reading it from the Data file.

The way this function is supposed to be used is: if it returns a key, good — you can use the key returned by the index and avoid reading it from the Data file. If it doesn't return a key — too bad, you have to read it from the Data file.

I'll add a code comment about this.

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -1622,9 +1649,55 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+                            }
+                        } else {
+                            // Case 2 from the comment above.
+                            return advance_index_until_unseen_partition().then([this] {
```
Let's say that there are 4 partitions, with keys `aa`, `bb`, `cc`, `dd`.
The index will store prefixes `a`, `b`, `c`, `d`.

Imagine that the reader is forwarded to range `(-∞, bc]` first, and to `[bd, +∞)` later. And imagine that the read is slicing, so that the data reader finishes reading partition `bb` somewhere in its middle.

Since the index doesn't know whether the key corresponding to prefix `b` is before `bc`/`bd` or after them, it has to be conservative and include `b` inside both ranges. So it will return the data range corresponding to `(-∞, b]` for the first forward, and `[b, +∞)` for the second forward.

The second forward happens after the data reader is done with partition `bb`, and is positioned somewhere in the middle of it. For the second forward, the index reader returns `bb` as the lower bound, which is before the current data reader position.

So what do we do in this situation?

If the data reader could move backwards, we could just set the data reader to `bb` again and let it be skipped by the piece of logic which skips keys before current range. But the whole problem is that the data reader can't move backwards. We can only move forward. So that's what we do — we find the next partition (i.e. the first one after current data reader position) and we skip there.

Note that the logic is written in a way which allows the reader to be more inexact than off-by-one, but with BTI, the `advance_index_until_unseen_partition` will always advance by exactly one partition.

> Does index_reader::advance_to_next_partition() have better granularity than index_reader::advance_to()?

It advances by exactly one partition from where the index lower bound currently is. It's not about granularity — both methods have full granularity, because the index does store the offsets of every partition. But for some range queries, BTI has no way to know which one of two entries is the *first*(/*last*) one inside the range, so it has to make the conservative choice, which on each end of the range potentially includes one extra partition.

michoecho commented:
I guess this is ready for queueing, but the patch is complicated (and so — possibly controversial) enough that we could give other people (e.g. @denesb?) at least a chance to look at it first.

nyh reviewed (commented):
The patch looks good to me, but I don't feel familiar enough with this code to actually pass an approve/disapprove judgement so I'll leave that to others. I made a few comments, none of them important.

I would reconsider the mention of the word "BTI" throughout the code and the commit messages - even after finding the meaning of it in Cassandra's site, it still doesn't make sense how "Trie-indexed Based SSTable" gets an acronym BTI. Do we have to use this acronym? :-)

denesb reviewed (changes requested):
I found one small bug (possible crash with trace logging), LGTM otherwise.

michoecho commented on test/boost/sstable_inexact_index_test.cc:
```diff
@@ -0,0 +1,607 @@
+    std::vector<uint64_t> result;
+    auto abstract_r = sst->make_index_reader(permit, nullptr, use_caching::no, false);
+    auto& r = dynamic_cast<index_reader&>(*abstract_r);
+    r.advance_to(dht::partition_range::make_open_ended_both_sides()).get();
```
Different overload. The removed one was `advance_to(ring_position_view)`. This is `advance_to(partition_range)`.

michoecho commented on tools/scylla-sstable.cc:
```diff
@@ -1081,8 +1081,10 @@ void dump_index_operation(schema_ptr schema, reader_permit permit, const std::ve
             writer.StartObject();
-            writer.Key("key");
-            writer.DataKey(*schema, pkey);
+            if (pkey) {
```
True.

avikivity commented on sstables/index_reader.hh:
```diff
@@ -125,9 +125,9 @@ public:
     // Precondition: partition_data_ready()
     virtual std::optional<sstables::deletion_time> partition_tombstone() = 0;
-    // Returns the key for current partition.
+    // Returns the key for current partition, if available in the index.
```
Multithreading is hard.

Possibly it referred to something else when it was a datastax enterprise feature, then retrofitted.

michoecho replied:
Refer to https://github.com/apache/cassandra/blob/cassandra-5.0/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md

BTI stands for "big trie-indexed".

Why the sstables are called "big", I don't know, I'm not an archaeologist.

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -1622,9 +1649,55 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+                        // 2. The parser was not advanced past the last range, but the inexact
+                        // index returned a start position for the new range which in reality
+                        // lies inside the old range. In this case, we should start reading
+                        // from the first partition with position greater or equal to the parser's position.
```
I'm not sure how to detect that.

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -1622,9 +1649,55 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
                     parse_assert(bool(end), _sst->get_filename());
+                    sstlog.trace("mp_row_consumer_reader_mx {}: fast_forward_to({}), index returned range [{}, {}), parser currently at {}", fmt::ptr(this), _pr, start, *end, _context->position());
+                    if (start < _context->position()) {
+                        sstlog.trace("mp_row_consumer_reader_mx {}: _paused_after_partition_start={}", fmt::ptr(this), _paused_after_partition_start);
```
> Please consider if they are really free (or almost free), and if you think they'll ever be helpful in the future.

First of all, avoiding trace logs because of performance impact is a tragedy. Scylla needs more trace logs, not less. We should be spamming trace logs everywhere. Debugging Scylla during development is already hard enough.
If you ask me, I think that we should be compiling trace logs away in release builds, because they are only useful in developer builds anyway. I'd gladly make this a devbuild-only log, if we had a convention for that. But we don't. So `trace` logs are the next best thing.

Second, I do keep the cost in mind when I write those logs. (Which is a bad thing, I shouldn't have to do that). I can infer from the source code that it should be almost free. A silent log (which doesn't require any computation to prepare the arguments) should compile to a read of a hot variable (the log level) and a predictable branch on that variable. And the logging code should be placed out of the hot path by PGO.
It adds some constraints on the optimizer, but they shouldn't really matter, it's not like they are in hot loops.
It also wastes branch predictor slots. And that's pretty much it.

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -33,7 +33,7 @@ class mp_row_consumer_reader_mx : public mp_row_consumer_reader_base, public mut
     }
 
-    void on_next_partition(dht::decorated_key, tombstone);
+    virtual data_consumer::proceed on_next_partition(dht::decorated_key, tombstone);
```
>Can you please explain (also in the commit message) why this method became virtual - will different classes implement it differently? How/why?

This method is used by two subclasses. `mx_sstable_full_scan_reader` (which exists as a fallback for sstables without index, and isn't used during normal database operation) and `mx_sstable_mutation_reader` (the main one).

In this patch, I made this method virtual to override the logic of `on_next_partition` for `mx_sstable_mutation_reader`. I didn't want to touch `mx_sstable_full_scan_reader` at all, and it retained the old implementation of the method.

> By the way, did you measure the performance implications of this patch with the original index format? I presume there is some small overhead, but how small is it?

I didn't measure the performance implications, because I don't see anything in this patch which could have non-negligible implication. But I can find and run some sstable benchmark, if you insist.

(And by the way, we shouldn't have to talk about that. We should have a CI that can notice performance regressions in the hot path).

nyh replied:
Ok. I'll trust your intuition. I was a bit worried that this is a very low level function and making it virtual makes it slightly slower. But on the other hand, there is so much other code involved here - and, heck, I/O, that you're probably right that there's no chance that the extra virtual call will make any measurable difference. So I'm not expecting you to do any specific measurements.

I agree with you about CI, but the reality is that 1. we don't have it, and 2. even when we do have it, it's not obvious how to bisect regressions so we usually don't do it (see today's thread with Israel and Avi discussing some write path regression).

michoecho commented:
>I would reconsider the mention of the word "BTI" throughout the code and the commit messages - even after finding the meaning of it in Cassandra's site, it still doesn't make sense how "Trie-indexed Based SSTable" gets an acronym BTI. Do we have to use this acronym? :-)

@nyh It's basically a proper noun. It doesn't matter what it stands for. And in the absence of a better idea it makes sense to use the one that is already published. Do you have a better name?

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -2250,15 +2260,21 @@ future<uint64_t> validate(
+                if (pk) {
+                    sstlog.trace("validate(): index-data position check for partition {}: {} == {}", *pk, data_pos, index_pos);
+                } else {
+                    sstlog.trace("validate(): index-data position check for a partition: {} == {}", *pk, data_pos, index_pos);
```
There have been many issues about this opened over the years, but the author of the library closes them because he considers it a feature, not a bug. (And refers people https://github.com/fmtlib/fmt/issues/492 for the rationale).

michoecho commented:
>  I think you need to explain this in the comment. Reading the current comment, it's not clear how one can ever use a function that may or may not return a response, for unexplained reasons. It returns the key but only "if available in the index" - without knowing when it's available and when it isn't. You're explaining that in cases where Scylla knows it will need this key, it makes sure it's written, but maybe you can refer to that additional code that makes sure of that, so the reader knows where this needs to be ensured?
>
> I wonder why we didn't just add the full key for every index location - it seems not having the full key complicated everything (and necessitated this pull request). Is the reason for that wishing to support use cases with very long keys?

@nyh `get_partition_key()` (i.e. storing a full partition key in the index) is an optimization which allows the data reader to avoid reading the partition header from the Data file just to get the key, in cases where the queried partition slice is potentially far away from the header on the disk.

The reader shouldn't rely on the full key being present. That's an arbitrary choice of the writer of the index. If the full key is not present, the reader should just fall back to reading it from the Data file.

The way this function is supposed to be used is: if it returns a key, good — you can use the key returned by the index and avoid reading it from the Data file. If it doesn't return a key — too bad, you have to read it from the Data file.

michoecho commented on sstables/mx/reader.cc:
```diff
@@ -1719,6 +1722,44 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+            // A BTI index is inexact in general (it can return a Data file range
+            // which includes some partitions before and after the queried range),
+            // but it is guaranteed to return the exact position if it's queried
+            // for a key which is present in the sstable.
```
By construction, BTI indexes store a prefix that is long enough to distinguish a present key from its neighbours. (I.e. up to — and including — the first byte which distinguishes from the neighbours).

If both keys aaaaaaaaaaaaaaaaaaaaa and aaaaaaaaaaaaaaaaaaaab exist, then the trie will look like
```
a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a - a
                                        \ b
```

And during the query, the index reader will walk it until the leaf (i.e. the `b`).

nyh replied:
I feel that I'm still missing something, about when a key can be missing in the index (I thought you said such a case), but clearly you do understand what you're talking about, so I'll trust your understanding and not waste your time trying to ask you all the details.

michoecho commented:
v1.02:
- Fixed the bad optional access in a trace log in `validate`.
- Added an explanatory comment to `abstract_index_reader::get_partition_key`.
- Applied the cosmetic suggestions.

nyh commented:
> > I would reconsider the mention of the word "BTI" throughout the code and the commit messages - even after finding the meaning of it in Cassandra's site, it still doesn't make sense how "Trie-indexed Based SSTable" gets an acronym BTI. Do we have to use this acronym? :-)
> 
> @nyh It's basically a proper noun. It doesn't matter what it stands for. And in the absence of a better idea it makes sense to use the one that is already published. Do you have a better name?

The name doesn't matter much, but introducing for the first time without explaining it does matter. Most Scylla developers and users are cutting-edge Cassandra users and would have no clue what "BTI" means. I had no idea what it means - I thought maybe it's some sort of famous computer science algorithm that for some reason I missed, but searches yielded nothing (before I thought of adding "Cassandra" to my searches and I finally got results).
I saw in another patch you referred to it as "a trie-based index format (The same or almost the same as Cassandra's BTI).". I think that's clearer than just assuming that users know what BTI means.

nyh commented on sstables/mx/reader.cc:
```diff
@@ -1719,6 +1722,44 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+
+    data_consumer::proceed on_next_partition(dht::decorated_key key, tombstone tomb) override {
+        if (_pr.get().before(key, dht::ring_position_comparator(*_schema))) {
+            sstlog.trace("mp_row_consumer_reader_mx {}: on_next_partition({}), _pr={}, skipping key before range", fmt::ptr(this), key, _pr);
```
Ok, it seems to me that these traces are unusable to anyone except the person who originally wrote the code - and this person can add his own printouts when debugging - but I'll trust your intuition on this.

denesb commented:
Dequeued, breaks promotion.
Looks like CI only ran modified tests, this is the second such case I see this week.

michoecho commented:
> 🔴 CI State: FAILURE

Seems like some podman issue on `spider6`.

avikivity commented:
@yaronkaikov is the test result above full or partial?

yaronkaikov commented:
> @yaronkaikov is the test result above full or partial?

you can use the link to find out, based on the timeline i assume it's full. since the fix was merged on Friday

yaronkaikov commented:
> > > @yaronkaikov is the test result above full or partial?
> > 
> > 
> > you can use the link to find out, based on the timeline i assume it's full. since the fix was merged on Friday
> 
> There are > 10 links there.
> 
> I clicked on test.py randomly, but the allure report for that is empty.

Well, i don't know why  (and i am not using this Allure :-) ) , but if you click on the link (any of time) and go to testReport you will see the info - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18545/testReport/

mykaul commented:
> Dequeued, breaks promotion. Looks like CI only ran modified tests, this is the second such case I see this week.

@michoecho - what were the changes since? Can this now be re-reviewed for inclusion?

michoecho commented:
> @avikivity - please re-review.

There's nothing to re-review. The only change since last time is a one-liner fix. (I.e. `auto _last_key = ...` fixed to `_last_key = ...`.

Please re-queue it, though.

nyh commented:
Created a pull request to fix the build on older fmt: https://github.com/scylladb/scylladb/pull/25228
